src/transaction/log_checkpoint_info.cpp

```diff
@@ -143,32 +151,135 @@ namespace cublog
+  void
+  checkpoint_info::load_checkpoint_trans (log_tdes &tdes, LOG_LSA &smallest_lsa)
   {
+    if (tdes.trid != NULL_TRANID && !LSA_ISNULL (&tdes.tail_lsa))
+      {
+	m_trans.emplace_back ();
+	LOG_INFO_CHKPT_TRANS &chkpt_tran = m_trans.back ();
+
+	chkpt_tran.isloose_end = tdes.isloose_end;
+	chkpt_tran.trid = tdes.trid;
+	chkpt_tran.state = tdes.state;
+
+	LSA_COPY (&chkpt_tran.head_lsa, &tdes.head_lsa);
+	LSA_COPY (&chkpt_tran.tail_lsa, &tdes.tail_lsa);
+	if (chkpt_tran.state == TRAN_UNACTIVE_ABORTED)
+	  {
+	    /*
+	     * Transaction is in the middle of an abort, since rollback does
+	     * is not run in a critical section. Set the undo point to be the
+	     * same as its tail. The recovery process will read the last
+	     * record which is likely a compensating one, and find where to
+	     * continue a rollback operation.
+	     */
+	    LSA_COPY (&chkpt_tran.undo_nxlsa, &tdes.tail_lsa);
+	  }
+	else
+	  {
+	    LSA_COPY (&chkpt_tran.undo_nxlsa, &tdes.undo_nxlsa);
+	  }
+
+	LSA_COPY (&chkpt_tran.posp_nxlsa, &tdes.posp_nxlsa);
+	LSA_COPY (&chkpt_tran.savept_lsa, &tdes.savept_lsa);
+	LSA_COPY (&chkpt_tran.tail_topresult_lsa, &tdes.tail_topresult_lsa);
+	LSA_COPY (&chkpt_tran.start_postpone_lsa, &tdes.rcv.tran_start_postpone_lsa);
+	std::strncpy (chkpt_tran.user_name, tdes.client.get_db_user (), LOG_USERNAME_MAX);
+
+	if (LSA_ISNULL (&smallest_lsa) || LSA_GT (&smallest_lsa, &tdes.head_lsa))
+	  {
+	    LSA_COPY (&smallest_lsa, &tdes.head_lsa);
+	  }
+      }
+  }
+
+  void
+  checkpoint_info::load_checkpoint_topop (log_tdes &tdes)
+  {
+    if (tdes.trid != NULL_TRANID && (!LSA_ISNULL (&tdes.rcv.sysop_start_postpone_lsa)
+				     || !LSA_ISNULL (&tdes.rcv.atomic_sysop_start_lsa)))
+      {
+	/* this transaction is running system operation postpone or an atomic system operation
+	 * note: we cannot compare tdes->state with TRAN_UNACTIVE_TOPOPE_COMMITTED_WITH_POSTPONE. we are
+	 *       not synchronizing setting transaction state.
+	 *       however, setting tdes->rcv.sysop_start_postpone_lsa is protected by
+	 *       log_Gl.prior_info.prior_lsa_mutex. so we check this instead of state.
+	 */
+	m_sysops.emplace_back();
+	LOG_INFO_CHKPT_SYSOP &chkpt_topop = m_sysops.back();
+	chkpt_topop.trid = tdes.trid;
+	chkpt_topop.sysop_start_postpone_lsa = tdes.rcv.sysop_start_postpone_lsa;
+	chkpt_topop.atomic_sysop_start_lsa = tdes.rcv.atomic_sysop_start_lsa;
+      }
+  }
+
+  void
+  checkpoint_info::load_trantable_snapshot (THREAD_ENTRY *thread_p, LOG_LSA &smallest_lsa)
+  {
+
+    LOG_TDES *act_tdes;		/* Transaction descriptor of an active transaction */
+    int i;
+    log_system_tdes::map_func mapper;
+
+    //ENTER the critical section
+    TR_TABLE_CS_ENTER (thread_p);
+    log_Gl.prior_info.prior_lsa_mutex.lock ();
+
+    scope_exit <std::function<void (void)>> unlock_on_exit ([thread_p] ()
+    {
+      log_Gl.prior_info.prior_lsa_mutex.lock ();
+      TR_TABLE_CS_EXIT (thread_p);
+    });
```

Comment:
Does it work like this?
```suggestion
    scope_exit unlock_on_exit ([thread_p] ()
    {
      log_Gl.prior_info.prior_lsa_mutex.lock ();
      TR_TABLE_CS_EXIT (thread_p);
    });
```